keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,1048 @@
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- Scheduler doesn't actively clear a Pod's `schedulingGates` field. This means if some controller
+is ill-implemented, some Pods may stay in Pending state incorrectly. If you noticed a Pod stays
+in Pending state for a long time and it carries non-nil `schedulingGates`, you may find out which
+component owns the gate(s) (via `.spec.managedFields`) and report the symptom to the component owner.
+
+- Faulty controllers may forget to remove the Pod's `schedulingGates`, and hence results in
+a large number of unschedulable Pods. In Alpha, we don't limit the number of unschedulabe Pods caused
+by potential faulty controllers. We will evaluate necessary options in the future to mitigate
+potential abuse.
+
+- End-users may be confused by no scheduling events in the output of `kubectl describe pod xyz`. We
+will provide detailed documentation, along with metrics, tooling (kubectl) and(or) events.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+A new API field `SchedulingGates` will be added to Pod's spec:
+
+```go
+type PodSpec struct {
+    // Each scheduling gate represents a particular scenario the scheduling is blocked upon.
+    // Scheduling is triggered only when SchedulingGates is empty.
+    // In the future, we may impose permission mechanics to restrict which controller can mutate
+    // which schedulingGate. It's dependent on a yet-to-be-implemented fined-grained
+    // permission (https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)
+    // and needs to be consistent with how finalizes/liens work today.
+    SchedulingGates []PodSchedulingGate
+}
+
+type PodSchedulingGate struct {
+    // Name of the SchedulingGate.
+    Name string
+}
+```
+
+In the scheduler's ComponentConfig API, we'll add a new type of extension point called `Enqueue`:
+
+```go
+type Plugins struct {
+    ......
+    Enqueue PluginSet
+}
+```
+
+### Implementation
+
+Inside scheduler, an internal queue called "activeQ" is designed to store all ready-to-schedule Pods.
+In this KEP, we're going to wire the aforementioned `Enqueue` extension to the logic deciding whether
+or not to add Pods to "activeQ".
+
+Specifically, prior to adding a Pod to "activeQ", scheduler iterates over registered Enqueue plugins.
+An Enqueue plugin must implement the `EnqueuePlugin` interface to return a `Status` to tell scheduler
+whether this Pod can be admitted or not:
+
+```go
+// EnqueuePlugin is an interface that must be implemented by "EnqueuePlugin" plugins.
+// These plugins are called prior to adding Pods to activeQ.
```

Comment:
we should emphasize that those plugins need to be lightweight because they will be invoked from event handlers.

Reply:
updated.